sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/json/JsonInferSchema.scala

```diff
@@ -121,7 +122,26 @@ private[sql] class JsonInferSchema(options: JSONOptions) extends Serializable {
             DecimalType(bigDecimal.precision, bigDecimal.scale)
         }
         decimalTry.getOrElse(StringType)
-      case VALUE_STRING => StringType
+      case VALUE_STRING =>
+        val stringValue = parser.getText
```

Comment:
shall we abstract out this logic for all the text sources?

Reply:
Yes, we can do that. There is some common code that could be shared. Can we do it in a separate PR?

Reply:
sure. How many text data sources already support it?

Reply:
`DateType` is not inferred at all but there is another type inference code that could be shared between JSON and CSV (maybe somewhere else).

Reply:
I checked `PartitioningUtils.inferPartitionColumnValue`, we try timestamp first and then date. Shall we follow it?

Reply:
or the order doesn't matter?

Reply:
The order can be matter if you have the same pattern (or similar) for dates and timestamps. `DateType` can be preferable because it requires less memory. 

It seems reasonable to move from `DateType` to `TimestampType` during schema inferring since opposite one is impossible without loosing info.

Reply:
If we switch the order here, we don't need the length check [here](https://github.com/apache/spark/pull/23201/files#diff-e925de14239f40430d05f9ffd0360f10R130), right?

Reply:
do you mean partition value type inference will have a different result than json value type inference?

Reply:
the partition feature is shared between all the file-based sources, I think it's an overkill to make it differ with different data sources.

The simplest solution to me is asking all text sources to follow the behavior of partition value type inference.

Reply:
Yea, one time I tried to match it with CSV a long long ago but I kind of gave up due to behaviour changes IIRC. If that's possible, it should be awesome.

If that's difficult, matching the behaviour within text based datasource (meaning CSV and JSON I guess) should be good enough.

Reply:
> If we switch the order here, we don't need the length check here, right?

@cloud-fan, that works only if we use default date/timestamp patterns. Both should do the exact match with pattern, which unfortunately the current parsing library (SimpleDateFormat) does not allow.

The order here is just to make it look better and both shouldn't be dependent on its order. I think we should support those inferences after completely switching the library to `java.time.format.*` without a legacy. That should make this change easier without a hole.